ironfish/src/blockchain/database/blockchaindb.ts

```diff
@@ -438,33 +438,34 @@ export class BlockchainDBTransaction implements IDatabaseTransaction {
   }
 
   async acquireLock(): Promise<void> {
-    if (!this.checkpoint) {
+    if (!this.stateRoot) {
+      this.stateRoot = await this.stateManager.getStateRoot()
       await this.stateManager.checkpoint()
-      this.checkpoint = true
     }
     await this.tx.acquireLock()
   }
 
   async update(): Promise<void> {
-    if (this.checkpoint) {
+    if (this.stateRoot) {
       await this.stateManager.commit()
-      this.checkpoint = false
+      this.stateRoot = null
     }
     await this.tx.update()
   }
 
   async commit(): Promise<void> {
-    if (this.checkpoint) {
+    if (this.stateRoot) {
       await this.stateManager.commit()
-      this.checkpoint = false
+      this.stateRoot = null
```

Comment:
Why set these to null instead of leaving them as the value? Is there a situation where the `stateRoot` will be different at the start of the next operation?

Reply:
Setting to null makes it so that we can't/won't roll back to the previous stateRoot after committing database changes. If we do something like this
```
withTransaction((tx) => {
  // add some values
  tx.update()
  
  // add some more values
  
  // throw an error
}
```
we want the stateRoot to include the first set of changes to be consistent with the other database stores